src/txn/txn_timestamp.c

```diff
@@ -630,166 +629,164 @@ __wt_txn_commit_timestamp_validate(WT_SESSION_IMPL *session, const char *name,
+		 */
+		if (has_oldest_ts && commit_ts < oldest_ts) {
+			__wt_timestamp_to_string(commit_ts, ts_string[0]);
+			__wt_timestamp_to_string(
+			    oldest_ts, ts_string[1]);
+			WT_RET_MSG(session, EINVAL,
+			    "commit timestamp %s is less than the oldest "
+			    "timestamp %s",
+			    ts_string[0], ts_string[1]);
+		}
 
-	/*
-	 * Compare against the prepare timestamp of the current transaction.
-	 * Return an error if the given timestamp is older than the prepare
-	 * timestamp.
-	 *
-	 * If roundup timestamps is configured, the commit timestamp will be
-	 * rounded up to the prepare timestamp.
-	 */
-	if (F_ISSET(txn, WT_TXN_PREPARE) &&
-	    !F_ISSET(txn, WT_TXN_TS_ROUND_PREPARED) &&
-	    ts < txn->prepare_timestamp) {
-		__wt_timestamp_to_string(txn->prepare_timestamp, ts_string[0]);
-		WT_RET_MSG(session, EINVAL,
-		    "%s timestamp %.*s older than the prepare "
-		    "timestamp %s for this transaction",
-		    name, (int)cval->len, cval->str, ts_string[0]);
-	}
+		if (has_stable_ts && commit_ts < stable_ts) {
+			__wt_timestamp_to_string(commit_ts, ts_string[0]);
+			__wt_timestamp_to_string(
+			    oldest_ts, ts_string[1]);
+			WT_RET_MSG(session, EINVAL,
+			    "commit timestamp %s is less than the stable "
+			    "timestamp %s",
+			    ts_string[0], ts_string[1]);
+		}
 
-	if (F_ISSET(txn, WT_TXN_HAS_TS_DURABLE) &&
-	    txn->durable_timestamp < txn->commit_timestamp) {
-		__wt_timestamp_to_string(txn->durable_timestamp, ts_string[0]);
-		__wt_timestamp_to_string(txn->commit_timestamp, ts_string[1]);
-		WT_RET_MSG(session, EINVAL,
-		    "%s timestamp %s older than the commit timestamp %s "
-		    "for this transaction",
-		    name, ts_string[0], ts_string[1]);
+		/*
+		 * Compare against the commit timestamp of the current
+		 * transaction.  Return an error if the given timestamp is
+		 * older than the first commit timestamp.
+		 */
+		if (F_ISSET(txn, WT_TXN_HAS_TS_COMMIT) &&
+		    commit_ts < txn->first_commit_timestamp) {
+			__wt_timestamp_to_string(commit_ts, ts_string[0]);
+			__wt_timestamp_to_string(
+			    txn->first_commit_timestamp, ts_string[1]);
+			WT_RET_MSG(session, EINVAL,
+			    "commit timestamp %s older than the first "
+			    "commit timestamp %s for this transaction",
+			    ts_string[0], ts_string[1]);
+		}
+	} else {
+		/*
+		 * For a prepared transaction, the commit timestamp should not
+		 * be less than the prepare timestamp.
+		 */
+		if (txn->prepare_timestamp > commit_ts) {
+			if (!F_ISSET(txn, WT_TXN_TS_ROUND_PREPARED)) {
+				__wt_timestamp_to_string(
+				    commit_ts, ts_string[0]);
+				__wt_timestamp_to_string(
+				    txn->prepare_timestamp, ts_string[1]);
+				WT_RET_MSG(session, EINVAL,
+				    "commit timestamp %s is less than the "
+				    "prepare timestamp %s for this transaction",
+				    ts_string[0], ts_string[1]);
+			}
+			commit_ts = txn->prepare_timestamp;
+		}
 	}
-
+	txn->durable_timestamp = txn->commit_timestamp = commit_ts;
```

Comment:
Could you add a `WT_ASSERT` above the assignment that either `txn->durable_timestamp == WT_TS_NONE` or `txn->durable_timestamp == txn->commit_timestamp`?

i.e: that we've never explicitly set the durable timestamp, before overriding it.